### PR TITLE
pkg/trace/writer: reduce sender count and improve retry

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This change reduces the default number of concurrent senders from 120 to 10

This change also improves the retry mechanism by having a per-payload backoff, and retrying payloads as they fail, rather than attempting to requeue them, which can result in dropping retryable payloads when the queue is full.

### Motivation

Having 120 simultaneous senders means that 120 payloads can be in-flight at a given time, which is far too high considering the amount of memory it can require (120 * ~10MiB = ~1.2GiB)

An I/O-bound operation like sending a payload does not benefit from high concurrency, so reducing this number should not affect throughput.

Retrying be re-queueing can result in dropping retryable payloads  when the send queue is full, and sender-wide backoff results in slowing down sends on all payloads when any failures are encountered, which is not desirable.

### Describe how to test/QA your changes

This will be tested in the various benchmarking platforms and on staging.

### CODE REVIEW

File: pkg/trace/writer/sender.go [hunk 1]
```diff
@@ -234,10 +230,22 @@ func (s *sender) Push(p *payload) {
...
 func (s *sender) sendPayload(p *payload) {
+	for attempt := 0; ; attempt++ {
+		s.backoff(attempt)
+		if s.sendOnce(p) {
+			return
+		}
+	}
+}
+
+// sendOnce attepmts to send the payload one time, returning
```
Comment (ajgajg1134) [inline pkg/trace/writer/sender.go:241]:
```suggestion
// sendOnce attempts to send the payload one time, returning
```

File: pkg/trace/writer/trace.go [hunk 2]
```diff
@@ -110,7 +110,7 @@ func NewTraceWriter(
 	}
 	climit := cfg.TraceWriter.ConnectionLimit
 	if climit == 0 {
-		climit = 100
+		climit = 5
```
Comment (ajgajg1134) [inline pkg/trace/writer/trace.go:113]: just an idea here: would it be easier to test this by just changing the config in staging to start and seeing how that affects performance there? If not then this is fine!

**Review by ajgajg1134 — APPROVED** [review]

File: pkg/trace/writer/trace.go (hunk 2 above)
Comment (pierotibou) [inline pkg/trace/writer/trace.go:113]: We lose the 4 to 1 ratio between stats and traces, is that a risk?

**Review by pierotibou — COMMENTED** [review]
> This change reduces the default number of concurrent senders from 120 to

In the description, the sentence is unfinished. I assume you meant 10. 

> than attempting to requeue them, which can result in dropping retryable payloads when the queue is full.

I'm fine with the new implementation but the old one was fine, no? I mean, dropping payload happens all the time arguably. Any other reason retrying on site is better than requeuing? Also, does the retry actually blocks one sender? Are there any potential side effect to retry instead of requeuing (I don't see any, just asking in case I would be missing one)?

File: pkg/trace/writer/trace.go (hunk 2 above)
Comment (ajgajg1134) [reply pkg/trace/writer/trace.go:114]: There's a small risk here, but I think with benchmarking / testing we can confirm if there are any issues and make adjustments before going to `main`

**Comment by knusbaum** [comment]
@pierotibou 
> I'm fine with the new implementation but the old one was fine, no? I mean, dropping payload happens all the time arguably. Any other reason retrying on site is better than requeuing? Also, does the retry actually blocks one sender? Are there any potential side effect to retry instead of requeuing (I don't see any, just asking in case I would be missing one)?


Just realized this wasn't addressed before we merged.

I think the old one was OK. It's just that under load, if we start getting retryable errors from intake for whatever reason, we're going to start throwing away a lot of payloads that *could* be retried without retrying them, because that queue is going to get full. This basically trades throughput for data loss. We'll have more backpressure because traces waiting in the queue won't get picked up right away, but we won't be dropping data unnecessarily. I think that's a good tradeoff, since we should always prefer not to drop unless we have no other choice. Eventually that backpressure will make it to our receiver and we'll start dropping there, but we'll report a 429 to the Tracers and they can make decisions about how much to retry.

**Pull request merged**